### PR TITLE
[SYCL] [FPGA] Add mutual diagnostic of max_concurrency attribute in conjunction of disable_loop_pipelining attribute

### PR DESCRIPTION
**Pull request by smanna12**
The support for disable_loop_pipelining was implemented on https://github.com/intel/llvm/pull/3441

This patch
  1. removes FIXME and adds mutual diagnostic of max_concurrency and disable_loop_pipelining attributes.
  2. updates document
  3. adds tests

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3320]: Happy days are here, we don't have to do this sort of dance any longer! You can remove this code (and some below) and instead do this all from Attr.td.

Below whichever attribute is defined last between disable_loop_pipelining and max_concurrency, add a definition to Attr.td like:
```
def : MutualExclusions<[SYCLIntelFPGAMaxConcurrency, SYCLIntelFPGADisableLoopPipelining]>;
```
Tablegen will then provide the checking logic for you so you don't have to explicitly add code for it. The only time we need to use `checkAttrMutualExclusion()` for declaration attributes is when the mutual exclusion has other logic to it (like ignoring implicitly-added attributes).

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:6380]: You can remove this code entirely as well.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:6419]: and this code too

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3320]: This would not work with the patch: 

https://github.com/intel/llvm/pull/3507/

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3320]: Oh, I see -- it's because of the `DeclOrStmtAttr` base class, that makes sense. Thanks!

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3320]: > Oh, I see -- it's because of the `DeclOrStmtAttr` base class, that makes sense. Thanks!

Yes.

Comment (premanandrao) [reply clang/lib/Sema/SemaDeclAttr.cpp:3320]: > I will hold this PR now until #3507 gets merged.

Yes, please, I had one (possibly minor) question there and it should get approved soon.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/include/clang/Basic/AttrDocs.td
Comment (smanna12) [reply clang/include/clang/Basic/AttrDocs.td]: Done

**Comment by smanna12** [comment]
I am seeing some problems with the merge, Please do not review.

**Comment by AaronBallman** [comment]
> I am seeing some problems with the merge, Please do not review.

I'm looking at those errors as well as they seem to be related to the recent tablegen changes.

**Comment by smanna12** [comment]
> > I am seeing some problems with the merge, Please do not review.
> 
> I'm looking at those errors as well as they seem to be related to the recent tablegen changes.

Yes, many warnings are generating like this:
 warning: ‘virtual bool clang::ParsedAttrInfo::diagMutualExclusion(clang::Sema&, const clang::ParsedAttr&, const clang::Stmt*) const’ was hidden [-Woverloaded-virtual]
  virtual bool diagMutualExclusion(Sema &S, const ParsedAttr &A,

@AaronBallman, are you seeing same?

File: clang/include/clang/Basic/Attr.td
Comment (AaronBallman) [inline clang/include/clang/Basic/Attr.td:1956]: Oh, this is the same as two lines up! I think this explains the build failures.

**Comment by AaronBallman** [comment]
> > > I am seeing some problems with the merge, Please do not review.
> > 
> > 
> > I'm looking at those errors as well as they seem to be related to the recent tablegen changes.
> 
> Yes, many warnings are generating like this:
> warning: ‘virtual bool clang::ParsedAttrInfo::diagMutualExclusion(clang::Sema&, const clang::ParsedAttr&, const clang::Stmt*) const’ was hidden [-Woverloaded-virtual]
> virtual bool diagMutualExclusion(Sema &S, const ParsedAttr &A,
> 
> @AaronBallman, are you seeing same?

The warnings are fixed upstream by: https://github.com/llvm/llvm-project/commit/2458aa0b9136e7616f529b027d1d478cf699340f#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b

**Comment by smanna12** [comment]
> > > > I am seeing some problems with the merge, Please do not review.
> > > 
> > > 
> > > I'm looking at those errors as well as they seem to be related to the recent tablegen changes.
> > 
> > 
> > Yes, many warnings are generating like this:
> > warning: ‘virtual bool clang::ParsedAttrInfo::diagMutualExclusion(clang::Sema&, const clang::ParsedAttr&, const clang::Stmt*) const’ was hidden [-Woverloaded-virtual]
> > virtual bool diagMutualExclusion(Sema &S, const ParsedAttr &A,
> > @AaronBallman, are you seeing same?
> 
> The warnings are fixed upstream by: [llvm/llvm-project@2458aa0#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b](https://github.com/llvm/llvm-project/commit/2458aa0b9136e7616f529b027d1d478cf699340f#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b)

I need to do rebase to be up with upstream.

**Comment by AaronBallman** [comment]
> > > > > I am seeing some problems with the merge, Please do not review.
> > > > 
> > > > 
> > > > I'm looking at those errors as well as they seem to be related to the recent tablegen changes.
> > > 
> > > 
> > > Yes, many warnings are generating like this:
> > > warning: ‘virtual bool clang::ParsedAttrInfo::diagMutualExclusion(clang::Sema&, const clang::ParsedAttr&, const clang::Stmt*) const’ was hidden [-Woverloaded-virtual]
> > > virtual bool diagMutualExclusion(Sema &S, const ParsedAttr &A,
> > > @AaronBallman, are you seeing same?
> > 
> > 
> > The warnings are fixed upstream by: [llvm/llvm-project@2458aa0#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b](https://github.com/llvm/llvm-project/commit/2458aa0b9136e7616f529b027d1d478cf699340f#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b)
> 
> I need to do rebase to be up with upstream.

Errr, it'd help if I linked the correct review -- this one fixed those issues: https://github.com/llvm/llvm-project/commit/65c22acfa4a412066e47c3171ff26fcbd62f970e#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b

**Comment by smanna12** [comment]
> > > > > > I am seeing some problems with the merge, Please do not review.
> > > > > 
> > > > > 
> > > > > I'm looking at those errors as well as they seem to be related to the recent tablegen changes.
> > > > 
> > > > 
> > > > Yes, many warnings are generating like this:
> > > > warning: ‘virtual bool clang::ParsedAttrInfo::diagMutualExclusion(clang::Sema&, const clang::ParsedAttr&, const clang::Stmt*) const’ was hidden [-Woverloaded-virtual]
> > > > virtual bool diagMutualExclusion(Sema &S, const ParsedAttr &A,
> > > > @AaronBallman, are you seeing same?
> > > 
> > > 
> > > The warnings are fixed upstream by: [llvm/llvm-project@2458aa0#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b](https://github.com/llvm/llvm-project/commit/2458aa0b9136e7616f529b027d1d478cf699340f#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b)
> > 
> > 
> > I need to do rebase to be up with upstream.
> 
> Errr, it'd help if I linked the correct review -- this one fixed those issues: [llvm/llvm-project@65c22ac#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b](https://github.com/llvm/llvm-project/commit/65c22acfa4a412066e47c3171ff26fcbd62f970e#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b)

Thanks @AaronBallman. I cherry-pick the change to fix all warnings about -Woverloaded-virtual.

File: clang/test/SemaSYCL/initiation_interval.cpp
Comment (AaronBallman) [inline clang/test/SemaSYCL/initiation_interval.cpp:6]: This seems like an unrelated change?

File: clang/include/clang/Basic/AttrDocs.td
Comment (AaronBallman) [inline clang/include/clang/Basic/AttrDocs.td:2838]: These are good changes, but they're unrelated to the patch (split them into an NFC change?).

File: clang/test/SemaSYCL/initiation_interval.cpp
Comment (AaronBallman) [inline clang/test/SemaSYCL/initiation_interval.cpp:32]: Also good changes but unrelated to the patch.

File: clang/utils/TableGen/ClangAttrEmitter.cpp
Comment (AaronBallman) [inline clang/utils/TableGen/ClangAttrEmitter.cpp:3699]: NB for other reviewers: this was a cherry pick from community and silences a bunch of warning diagnostics that were making it hard to read the CI pipeline output.

File: clang/include/clang/Basic/AttrDocs.td
Comment (smanna12) [reply clang/include/clang/Basic/AttrDocs.td]: Yes, Done.

File: clang/test/SemaSYCL/initiation_interval.cpp
Comment (smanna12) [reply clang/test/SemaSYCL/initiation_interval.cpp]: Done

File: clang/test/SemaSYCL/initiation_interval.cpp
Comment (smanna12) [reply clang/test/SemaSYCL/initiation_interval.cpp]: Yes, Done

File: clang/utils/TableGen/ClangAttrEmitter.cpp
Comment (smanna12) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:3699]: Cherry-pick from :  llvm/llvm-project@65c22ac#diff-58111c7c545672c3a67b1b31a37d4103b35d9c584a46c74b0be14cc37741491b

**Comment by smanna12** [comment]
buildbot/Lit_with_Cuda is failing with warning like this:

lit WARNING: The PID of hang binary not found

**Review by premanandrao — APPROVED** [review]
This looks good to me.

**Comment by bader** [comment]
> buildbot/Lit_with_Cuda is failing with warning like this:
> 
> lit WARNING: The PID of hang binary not found

@tfzhu, could you help to resolve that issue, please?

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by smanna12** [comment]
> > buildbot/Lit_with_Cuda is failing with warning like this:
> > lit WARNING: The PID of hang binary not found
> 
> @tfzhu, could you help to resolve that issue, please?

@tfzhu, any update on it? Thanks

**Comment by smanna12** [comment]
> > buildbot/Lit_with_Cuda is failing with warning like this:
> > lit WARNING: The PID of hang binary not found
> 
> @tfzhu, could you help to resolve that issue, please?

@bader, The issues are resolved. Thanks.

**Pull request merged**
